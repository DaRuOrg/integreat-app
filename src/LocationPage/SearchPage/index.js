import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import Layout from 'Layout/Layout'
import LANGUAGE_ENDPOINT from 'endpoints/language'

class SearchPage extends React.Component {
  static propTypes = {
    language: PropTypes.string.isRequired
  }

  componentWillUnmount () {
    this.props.dispatch(LANGUAGE_ENDPOINT.invalidateAction())
  }

  componentWillMount () {
    let location = this.getLocation()
    this.props.dispatch(LANGUAGE_ENDPOINT.fetchEndpointAction({
      location: location,
      language: this.props.language
    }))
  }

  getLocation () {
    return this.props.match.params.location
  }

  render () {
    return (
      <Layout currentLanguage={this.props.language}>
        TODO
      </Layout>
    )
  }
}

/**
 * @param state The current app state
 * @return {{locations: {}}}  The endpoint values from the state mapped to props
 */
function mapStateToProps (state) {
  return ({
    language: state.language.language
  })
}

export default connect(mapStateToProps)(SearchPage)
