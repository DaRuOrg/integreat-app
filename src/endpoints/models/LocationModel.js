import { isEmpty } from 'lodash/lang'

export default class LocationModel {
  constructor (name, path, live) {
    this._code = name
    this._name = path
    this._live = live
  }

  get live () {
    return this._live
  }

  get name () {
    return this._code
  }

  get path () {
    return this._name
  }

  get category () {
    return isEmpty(this._code) ? '?' : this._code[0].toUpperCase()
  }
}
