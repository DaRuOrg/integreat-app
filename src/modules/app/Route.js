import UrlPattern from 'url-pattern'

class Route {
  id
  path
  pattern

  constructor (id, path) {
    this.id = id
    this.path = path

    this.pattern = new UrlPattern(path)
  }

  stringify (params = {}) {
    return this.pattern.stringify(params)
  }

  hasPath (path) {
    return this.path === path
  }
}

export default Route
