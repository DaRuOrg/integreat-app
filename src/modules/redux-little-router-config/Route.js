import UrlPattern from 'url-pattern'

class Route {
  id
  path
  pattern
  renderComponent
  condition

  constructor (id, path, renderComponent, condition) {
    this.id = id
    this.path = path
    this.renderComponent = renderComponent
    this.condition = condition

    this.pattern = new UrlPattern(path)
  }

  stringify (params = {}) {
    return this.pattern.stringify(params)
  }
}

export default Route
