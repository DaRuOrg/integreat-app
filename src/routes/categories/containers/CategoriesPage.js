import React from 'react'
import PropTypes from 'prop-types'

import { connect } from 'react-redux'
import compose from 'lodash/fp/compose'

import Content from 'routes/categories/containers/Content'
import Breadcrumb from 'routes/categories/containers/Breadcrumb'
import Error from 'modules/common/containers/Error'
import PdfButton from 'routes/categories/components/PdfButton'
import withFetcher from 'modules/endpoint/hocs/withFetcher'
import PAGE_ENDPOINT from 'modules/endpoint/endpoints/pages'
import LANGUAGES_ENDPOINT from 'modules/endpoint/endpoints/languages'

import Hierarchy from '../Hierarchy'
import PageModel from 'modules/endpoint/models/PageModel'
import { setLanguageChangeUrls } from 'modules/language/actions/setLanguageChangeUrls'
import LanguageModel from 'modules/endpoint/models/LanguageModel'

/**
 * Matching the route /<location>/<language>*
 */
class CategoriesPage extends React.Component {
  static propTypes = {
    location: PropTypes.string.isRequired,
    languages: PropTypes.arrayOf(PropTypes.instanceOf(LanguageModel)).isRequired,
    language: PropTypes.string.isRequired,
    path: PropTypes.string,
    pages: PropTypes.instanceOf(PageModel).isRequired
  }

  getParentPath () {
    return `/${this.props.location}/${this.props.language}`
  }

  componentDidMount () {
    this.setLanguageChangeUrls(this.props.path)
  }

  // we must not call dispatch in componentWillUpdate or componentDidUpdate
  componentWillReceiveProps (nextProps) {
    if (nextProps.path !== this.props.path) {
      this.setLanguageChangeUrls(nextProps.path)
    }
  }

  mapLanguageToUrl = (language, id) => (
    id ? `/${this.props.location}/${language}/redirect?id=${id}` : `/${this.props.location}/${language}`
  )

  /**
   * Gets and stores the available languages for the current page
   * @param {string} path The current path
   */
  setLanguageChangeUrls (path) {
    const hierarchy = new Hierarchy(path)
    const error = hierarchy.build(this.props.pages)
    if (error) {
      // todo handle this error
      return
    }
    this.props.dispatch(setLanguageChangeUrls(
      this.mapLanguageToUrl, this.props.languages, hierarchy.top().availableLanguages)
    )
  }

  getPdfFetchPath () {
    let path = `/${this.props.location}/${this.props.language}/fetch-pdf/`
    if (this.props.path) {
      path += this.props.path
    }
    return path
  }

  render () {
    const url = this.getParentPath()
    const hierarchy = new Hierarchy(this.props.path)

    // Pass data to hierarchy
    const error = hierarchy.build(this.props.pages)
    if (error) {
      return <Error error={error}/>
    }

    return <div>
      <Breadcrumb
        hierarchy={hierarchy}
        language={this.props.language}
        location={this.props.location}
      />
      <Content url={url} hierarchy={hierarchy}/>
      <PdfButton href={this.getPdfFetchPath()}/>
    </div>
  }
}

const mapStateToProps = (state) => ({
  language: state.router.params.language,
  location: state.router.params.location,
  path: state.router.params['_'] // _ contains all the values from *
})

export default compose(
  connect(mapStateToProps),
  withFetcher(PAGE_ENDPOINT),
  withFetcher(LANGUAGES_ENDPOINT)
)(CategoriesPage)
