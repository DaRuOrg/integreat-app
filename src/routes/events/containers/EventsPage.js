import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import compose from 'lodash/fp/compose'
import Spinner from 'react-spinkit'

import EventModel from 'modules/endpoint/models/EventModel'
import EVENTS_ENDPOINT from 'modules/endpoint/endpoints/events'
import LANGUAGES_ENDPOINT from 'modules/endpoint/endpoints/languages'
import withFetcher from 'modules/endpoint/hocs/withFetcher'
import EventDetail from '../components/EventDetail'
import EventList from '../components/EventList'
import { setLanguageChangeUrls } from 'modules/language/actions/setLanguageChangeUrls'
import LanguageModel from 'modules/endpoint/models/LanguageModel'

/**
 * Displays a list of events or a single event, matching the route /<location>/<language>/events(/<id>)
 */
export class EventsPage extends React.Component {
  static propTypes = {
    events: PropTypes.arrayOf(PropTypes.instanceOf(EventModel)).isRequired,
    location: PropTypes.string.isRequired,
    languages: PropTypes.arrayOf(PropTypes.instanceOf(LanguageModel)).isRequired,
    language: PropTypes.string.isRequired,
    dispatchLanguageChangeUrls: PropTypes.func.isRequired,
    id: PropTypes.string
  }

  getPath () {
    return `/${this.props.location}/${this.props.language}/events`
  }

  mapLanguageToUrl = (language, id) =>
    id ? `/${this.props.location}/${language}/events/${id}`
      : `/${this.props.location}/${language}/events`

  findEvent = ((events, id) => events.find(
    (event) => event.id.toString() === id
  ))

  componentDidMount () {
    // all events
    let availableLanguages = {}

    if (this.props.id && this.props.events) {
      // specific event
      const event = this.findEvent(this.props.events, this.props.id)
      if (event) availableLanguages = event.availableLanguages
    }
    this.props.dispatchLanguageChangeUrls(this.mapLanguageToUrl, this.props.languages, availableLanguages)
  }

  // we must not call dispatch in componentWillUpdate or componentDidUpdate
  componentWillReceiveProps (nextProps) {
    if (nextProps.events === this.props.events && nextProps.id === this.props.id) {
      // no relevant prop changes
      return
    }

    if (nextProps.id) {
      // specific event
      const event = this.findEvent(nextProps.events, nextProps.id)
      if (event) {
        // events have been loaded in the new language
        this.props.dispatchLanguageChangeUrls(
          this.mapLanguageToUrl, nextProps.languages, event.availableLanguages
        )
      }
    } else {
      // all events
      this.props.dispatchLanguageChangeUrls(this.mapLanguageToUrl, nextProps.languages)
    }
  }

  render () {
    if (this.props.id) {
      // event with the given id from this.props.path
      const event = this.findEvent(this.props.events, this.props.id)

      if (event) {
        return <EventDetail event={event} location={this.props.location} language={this.props.language} />
      } else {
        // events in new language haven't been fetched yet
        return <Spinner name='line-scale-party' />
      }
    }
    return <EventList events={this.props.events} url={this.getPath()} language={this.props.language} />
  }
}

const mapStateToProps = (state) => ({
  language: state.router.params.language,
  location: state.router.params.location,
  id: state.router.params.id
})

const mapDispatchToProps = (dispatch) => ({
  dispatchLanguageChangeUrls: (urls, languages, availableLanguages) => dispatch(
    setLanguageChangeUrls(urls, languages, availableLanguages)
  )
})

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  withFetcher(EVENTS_ENDPOINT),
  withFetcher(LANGUAGES_ENDPOINT)
)(EventsPage)
