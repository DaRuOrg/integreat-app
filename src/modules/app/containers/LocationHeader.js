import React from 'react'
import PropTypes from 'prop-types'
import { translate } from 'react-i18next'
import { connect } from 'react-redux'
import compose from 'lodash/fp/compose'

import Navigation from 'modules/app/Navigation'
import LanguageSelector from './LanguageSelector'
import searchIcon from '../assets/magnifier.svg'
import locationIcon from '../assets/location-icon.svg'
import languageIcon from '../assets/language-icon.svg'
import LocationModel from 'modules/endpoint/models/LocationModel'
import withFetcher from 'modules/endpoint/hocs/withFetcher'
import LOCATIONS_ENDPOINT from 'modules/endpoint/endpoints/locations'
import Header from 'modules/layout/components/Header'

class LocationHeader extends React.Component {
  static propTypes = {
    navigation: PropTypes.instanceOf(Navigation).isRequired,
    location: PropTypes.instanceOf(LocationModel).isRequired,
    route: PropTypes.string.isRequired
  }

  isCategoriesEnabled = () => this.isExtrasEnabled() || this.isEventsEnabled()

  isExtrasEnabled = () => this.props.location.extrasEnabled

  isEventsEnabled = () => this.props.location.eventsEnabled

  isExtrasSelected = () => this.props.route === '/:location/:language/extras'

  isCategoriesSelected = () => ['/:location/:language', '/:location/:language/*'].includes(this.props.route)

  isEventsSelected = () => this.props.route === '/:location/:language/events(/:id)'

  getActionItems () {
    const {navigation} = this.props
    return [
      {href: navigation.search, iconSrc: searchIcon},
      {href: navigation.locationSelector, iconSrc: locationIcon},
      {dropDownNode: <LanguageSelector />, iconSrc: languageIcon}
    ]
  }

  getMenuItems () {
    const {t, navigation} = this.props
    const extras = this.isExtrasEnabled() &&
      {href: navigation.extras, active: this.isExtrasSelected(), text: t('extras')}

    const categories = this.isCategoriesEnabled() &&
      {href: navigation.categories, active: this.isCategoriesSelected(), text: t('categories')}

    const events = this.isEventsEnabled() &&
      {href: navigation.events, active: this.isEventsSelected(), text: t('news')}

    return [extras, categories, events].filter(item => item)
  }

  render () {
    return <Header actionItems={this.getActionItems()} menuItems={this.getMenuItems()} />
  }
}

const mapStateToProps = (state) => ({
  route: state.router.route,
  navigation: new Navigation(state.router.params.location, state.router.params.language)
})

export default compose(
  connect(mapStateToProps),
  withFetcher(LOCATIONS_ENDPOINT, true, true),
  translate('app')
)(LocationHeader)
