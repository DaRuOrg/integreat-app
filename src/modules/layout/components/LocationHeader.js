import React from 'react'
import PropTypes from 'prop-types'
import { translate } from 'react-i18next'

import LanguageSelector from '../../common/containers/LanguageSelector'
import searchIcon from '../assets/magnifier.svg'
import locationIcon from '../assets/location-icon.svg'
import languageIcon from '../assets/language-icon.svg'
import LocationModel from 'modules/endpoint/models/LocationModel'
import Header from 'modules/layout/components/Header'
import HeaderNavigationItem from '../HeaderNavigationItem'
import HeaderActionItem from '../HeaderActionItem'
import SearchPage from '../../../routes/search/containers/SearchPage'
import LandingPage from '../../../routes/landing/containers/LandingPage'
import CategoriesPage from '../../../routes/categories/containers/CategoriesPage'
import EventsPage from '../../../routes/events/containers/EventsPage'

class LocationHeader extends React.Component {
  static propTypes = {
    matchRoute: PropTypes.func.isRequired,
    currentParams: PropTypes.object.isRequired,
    location: PropTypes.instanceOf(LocationModel).isRequired,
    route: PropTypes.string.isRequired,
    t: PropTypes.func.isRequired
  }

  getActionItems () {
    const {matchRoute, currentParams} = this.props
    return [
      new HeaderActionItem({href: matchRoute(SearchPage).stringify(currentParams), iconSrc: searchIcon}),
      new HeaderActionItem({href: matchRoute(LandingPage).stringify(currentParams), iconSrc: locationIcon}),
      new HeaderActionItem({dropDownNode: <LanguageSelector />, iconSrc: languageIcon})
    ]
  }

  getNavigationItems () {
    const {t, matchRoute, currentParams, route} = this.props

    const isEventsEnabled = () => this.props.location.eventsEnabled
    const isExtrasEnabled = () => this.props.location.extrasEnabled
    const isCategoriesEnabled = () => isExtrasEnabled() || isEventsEnabled()

    const isExtrasSelected = () => false // todo for WEBAPP-64: test and verify this
    const isCategoriesSelected = () => matchRoute(CategoriesPage).hasPath(route)
    const isEventsSelected = () => matchRoute(EventsPage).hasPath(route)

    const extras = isExtrasEnabled() &&
      new HeaderNavigationItem({
        href: '/',
        active: isExtrasSelected(),
        text: t('extras')
      })

    const categories = isCategoriesEnabled() &&
      new HeaderNavigationItem({
        href: matchRoute(CategoriesPage).stringify(currentParams),
        active: isCategoriesSelected(),
        text: t('categories')
      })

    const events = isEventsEnabled() &&
      new HeaderNavigationItem({
        href: matchRoute(EventsPage).stringify(currentParams),
        active: isEventsSelected(),
        text: t('news')
      })

    return [extras, categories, events].filter(isEnabled => isEnabled)
  }

  render () {
    return <Header actionItems={this.getActionItems()} navigationItems={this.getNavigationItems()} />
  }
}

export default translate('app')(LocationHeader)
