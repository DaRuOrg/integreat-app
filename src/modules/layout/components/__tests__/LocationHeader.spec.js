import { shallow } from 'enzyme'
import React from 'react'
import LocationHeader from '../LocationHeader'
import LocationModel from 'modules/endpoint/models/LocationModel'
import Route from '../../../redux-little-router-config/Route'
import CategoriesPage from '../../../../routes/categories/containers/CategoriesPage'
import LandingPage from '../../../../routes/landing/containers/LandingPage'
import SearchPage from '../../../../routes/search/containers/SearchPage'
import EventsPage from '../../../../routes/events/containers/EventsPage'

describe('LocationHeader', () => {
  const matchRoute = (id) => {
    switch (id) {
      case CategoriesPage:
        return new Route(id, '/:location/:language(/*)')
      case LandingPage:
        return new Route(id, '/:language')
      case SearchPage:
        return new Route(id, '/:location/:language/search')
      case EventsPage:
        return new Route(id, '/:location/:language/events(/:id)')
    }
    throw new Error(`Route ${id} not found!`)
  }

  const currentParams = {
    location: 'augsburg',
    language: 'de'
  }

  const createLocation = (extrasEnabled, eventsEnabled) => new LocationModel({
    name: 'Mambo Nr. 5',
    code: 'location1',
    eventsEnabled,
    extrasEnabled
  })

  describe('NavigationItems', () => {
    test('should be empty, if extras and news are both disabled', () => {
      const component = shallow(<LocationHeader matchRoute={matchRoute} currentParams={currentParams}
                                                location={createLocation(false, false)}
                                                route='' />)
      expect(component.dive().prop('navigationItems')).toMatchSnapshot()
    })

    test('should show categories, if extras or news are enabled', () => {
      const extrasComp = shallow(<LocationHeader matchRoute={matchRoute} currentParams={currentParams}
                                                 location={createLocation(true, false)}
                                                 route='' />)
      const eventsComp = shallow(<LocationHeader matchRoute={matchRoute} currentParams={currentParams}
                                                 location={createLocation(false, true)}
                                                 route='' />)

      expect(extrasComp.dive().prop('navigationItems')).toMatchSnapshot()
      expect(eventsComp.dive().prop('navigationItems')).toMatchSnapshot()
    })

    test('should show extras, categories, events in this order', () => {
      const component = shallow(<LocationHeader matchRoute={matchRoute} currentParams={currentParams}
                                                location={createLocation(true, true)}
                                                route='' />)
      expect(component.dive().prop('navigationItems')).toMatchSnapshot()
    })

    test('should highlight categories if route corresponds', () => {
      const component = shallow(<LocationHeader matchRoute={matchRoute} currentParams={currentParams}
                                                      location={createLocation(true, true)}
                                                      route='/:location/:language(/*)' />)
      expect(component.dive().prop('navigationItems')).toMatchSnapshot()
    })

    test('should highlight events if route corresponds', () => {
      const component = shallow(<LocationHeader matchRoute={matchRoute} currentParams={currentParams}
                                                location={createLocation(true, true)}
                                                route='/:location/:language/events(/:id)' />)

      expect(component.dive().prop('navigationItems')).toMatchSnapshot()
    })

    // todo for WEBAPP-64: should highlight extras if route corresponds
  })

  test('should match snapshot', () => {
    const component = shallow(<LocationHeader matchRoute={matchRoute} currentParams={currentParams}
                                              location={createLocation(true, true)} route='' />)
    expect(component.dive()).toMatchSnapshot()
  })
})
