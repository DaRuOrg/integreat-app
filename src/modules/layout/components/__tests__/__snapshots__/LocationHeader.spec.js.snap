// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationHeader NavigationItems should be empty, if extras and news are both disabled 1`] = `Array []`;

exports[`LocationHeader NavigationItems should highlight categories if route corresponds 1`] = `
Array [
  HeaderNavigationItem {
    "_active": false,
    "_href": "/",
    "_text": "app:extras",
  },
  HeaderNavigationItem {
    "_active": true,
    "_href": "/augsburg/de",
    "_text": "app:categories",
  },
  HeaderNavigationItem {
    "_active": false,
    "_href": "/augsburg/de/events",
    "_text": "app:news",
  },
]
`;

exports[`LocationHeader NavigationItems should highlight events if route corresponds 1`] = `
Array [
  HeaderNavigationItem {
    "_active": false,
    "_href": "/",
    "_text": "app:extras",
  },
  HeaderNavigationItem {
    "_active": false,
    "_href": "/augsburg/de",
    "_text": "app:categories",
  },
  HeaderNavigationItem {
    "_active": true,
    "_href": "/augsburg/de/events",
    "_text": "app:news",
  },
]
`;

exports[`LocationHeader NavigationItems should show categories, if extras or news are enabled 1`] = `
Array [
  HeaderNavigationItem {
    "_active": false,
    "_href": "/",
    "_text": "app:extras",
  },
  HeaderNavigationItem {
    "_active": false,
    "_href": "/augsburg/de",
    "_text": "app:categories",
  },
]
`;

exports[`LocationHeader NavigationItems should show categories, if extras or news are enabled 2`] = `
Array [
  HeaderNavigationItem {
    "_active": false,
    "_href": "/augsburg/de",
    "_text": "app:categories",
  },
  HeaderNavigationItem {
    "_active": false,
    "_href": "/augsburg/de/events",
    "_text": "app:news",
  },
]
`;

exports[`LocationHeader NavigationItems should show extras, categories, events in this order 1`] = `
Array [
  HeaderNavigationItem {
    "_active": false,
    "_href": "/",
    "_text": "app:extras",
  },
  HeaderNavigationItem {
    "_active": false,
    "_href": "/augsburg/de",
    "_text": "app:categories",
  },
  HeaderNavigationItem {
    "_active": false,
    "_href": "/augsburg/de/events",
    "_text": "app:news",
  },
]
`;

exports[`LocationHeader should match snapshot 1`] = `
<Header
  actionItems={
    Array [
      HeaderActionItem {
        "_dropDownNode": undefined,
        "_href": "/augsburg/de/search",
        "_iconSrc": "test-file-stub",
      },
      HeaderActionItem {
        "_dropDownNode": undefined,
        "_href": "/de",
        "_iconSrc": "test-file-stub",
      },
      HeaderActionItem {
        "_dropDownNode": <Connect(getContext(Connect(languagesFetcher))) />,
        "_href": undefined,
        "_iconSrc": "test-file-stub",
      },
    ]
  }
  navigationItems={
    Array [
      HeaderNavigationItem {
        "_active": false,
        "_href": "/",
        "_text": "app:extras",
      },
      HeaderNavigationItem {
        "_active": false,
        "_href": "/augsburg/de",
        "_text": "app:categories",
      },
      HeaderNavigationItem {
        "_active": false,
        "_href": "/augsburg/de/events",
        "_text": "app:news",
      },
    ]
  }
/>
`;
